src/jit/optimizer.cpp

```diff
@@ -6142,9 +6142,15 @@ bool Compiler::optHoistLoopExprsForTree(GenTreePtr        tree,
         childrenCctorDependent[i] = false;
     }
 
-    // Initclass CLS_VARs are the base case of cctor dependent trees.
-    bool treeIsCctorDependent = (tree->OperIs(GT_CLS_VAR) && ((tree->gtFlags & GTF_CLS_VAR_INITCLASS) != 0));
-    bool treeIsInvariant      = true;
+    // Initclass CLS_VARs and IconHandles are the base cases of cctor dependent trees.
+    // In the IconHandle case, it's of course the dereference, rather than the constant itself, that is
+    // truly dependent on the cctor.  So a more precise approach would be to separately propagate
+    // isCctorDependent and isAddressWhoseDereferenceWouldBeCctorDependent, but we don't for simplicity/throughput;
+    // the constant itself would be considered non-hoistable anyway, since optIsCSEcandidate returns
+    // false for constants.
```

Comment:
Yes, GT_IND is already dipping into the common bits:

```c++
#define GTF_IND_NONFAULTING   0x00000800 // An indir that cannot fault.  GTF_SET_FLAGS is not used on indirs
```

We could (maybe?) similarly "steal" `GTF_ZSF_SET` or `GTF_UNSIGNED` for a new `GT_IND` flag, though I think the extra conservativism may be warranted here... with this change as-is, we'll block hoisting static field addresses produced by `ldsflda` (if the field is marked `CORINFO_FLG_FIELD_INITCLASS` and we aren't hoisting the cctor call with it).  The importer does the marking now, but the importer doesn't have the context to know when a `GT_IND` might be fed by a prior `ldsflda`, so it seems risky to leave the `ldsflda` addresses unmarked.  Using two bools here to track cctor-dependence would let us keep the annotations on the address and so avoid the flag sharing and the `ldsflda`+indir problem, I just didn't think it was worth it with no/little current benefit.  I could be persuaded to change it though, what do you think?